clients/hadoopfs/src/main/java/io/lakefs/FileSystemTracer.java

```diff
@@ -0,0 +1,353 @@
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.URI;
+
+/**
+ * This class implements a troubleshooting tool for the LakeFSFileSystem. You can use it to compare the output of file
+ * system operations done by {@link LakeFSFileSystem} and by {@link S3AFileSystem}.
+ *
+ * How to use the FileSystemTracer:
+ * To use the FileSystemTracer, you should set the value of the Hadoop configuration 'fs.lakefs.impl' to FileSystemTracer.
+ * This configures the FileSystemTracer to be the file system that handles paths with the lakefs scheme. i.e. paths with
+ * lakefs:// prefix.
+ *
+ * What does the FileSystemTracer do:
+ * The FileSystemTracer holds instances of {@link LakeFSFileSystem} and {@link S3AFileSystem}. On a file system operation,
+ * the FileSystemTracer invokes the operation on both file systems, logs the output of both calls, and returns the result
+ * of one of the file systems according to a configuration (S3AFileSystem output by default).
+
+ * Additional configurations:
+ * Mandatory -
+ * - fs.lakefs.tracer.working.dir - the s3 location in which the tracer can operate. This should be an S3 bucket name or
+ *   an absolute path of a directory on a bucket.
+ * Optional -
+ * - fs.lakefs.tracer.use.lakefs - tells the tracer whether it should return the response coming from the lakefs file
+ *   system or return s3a's response. by default it is set to false and returns s3a's response.
+ *
+ * Assumptions:
+ * - listing results of lakefs://repository/branch/$PATH and s3a://${fs.lakefs.tracer.working.dir}/$PATH are identical.
+ * - The s3 credentials available for Spark allow access to fs.lakefs.tracer.working.dir on s3.
+ */
+public class FileSystemTracer extends FileSystem {
+
+    public static final Logger LOG = LoggerFactory.getLogger(FileSystemTracer.class);
+    private static final String TRACER_WORKING_DIR = "fs.lakefs.tracer.working.dir";
+    private static final String USE_LAKEFS_RESPONSE = "fs.lakefs.tracer.use.lakefs";
+    private static final Object S3_URI_PREFIX = "s3";
+    private static final Object RESULTS_COMPARISON = "[RESULTS_COMPARISON]";
+
+
+    /*
+     A property that determines which file system's response the FileSystemTracer returns.
+     */
+    private boolean useLakeFSFileSystemRes;
+    private LakeFSFileSystem lfsFileSystem;
+    private FileSystem s3AFileSystem;
+    private String s3aPathPrefix;
+    private String lfsPathPrefix;
+
+    /**
+     * Transforms a lakefs path into an s3 path.
+     * Example:
+     *   in: lakefs://repository/branch/key=1/a.parquet
+     *   out: s3://${fs.lakefs.tracer.working.dir}/key=1/a.parquet
+     */
+    private Path translateLakeFSPathToS3APath(Path path) {
+        ObjectLocation objectLoc = lfsFileSystem.pathToObjectLocation(path);
+        String lakefsPath = path.toString();
+        String lakefsPrefix = String.format("%s://%s/%s", objectLoc.getScheme(), objectLoc.getRepository(),
```

Comment:
relevant for multiple places - we have a static method to format 3 and 4 parts addresses at ObjectLocation.formatPath